libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,482 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//
+//  Swift Navigation SBP GPS driver for ArduPilot.
+//	Code by Niels Joubert
+//
+//  Swift Binary Protocol format: http://docs.swift-nav.com/
+//
+
+#include "AP_GPS.h"
+#include "AP_GPS_SBP2.h"
+#include <DataFlash/DataFlash.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define SBP_DEBUGGING 1
+#define SBP_INFOREPORTING 1
+
+//INVARIANT: We expect SBP to give us a heartbeat in less than 2 seconds.
+//  This is more lax than the default Piksi settings,
+//  and we assume the user hasn't reconfigured their Piksi to longer heartbeat intervals
+#define SBP_TIMEOUT_HEARTBEAT  2000
+
+#if SBP_DEBUGGING
+ # define Debug(fmt, args ...)                  \
+do {                                            \
+    hal.console->printf("%s:%d: " fmt "\n",     \
+                        __FUNCTION__, __LINE__, \
+                        ## args);               \
+    hal.scheduler->delay(1);                    \
+} while(0)
+#else
+ # define Debug(fmt, args ...)
+#endif
+
+#if SBP_INFOREPORTING
+ # define Info(fmt, args ...)                  \
+do {                                            \
+    hal.console->printf("SBPv2: " fmt "\n",     \
+                        ## args);               \
+    hal.scheduler->delay(1);                    \
+} while(0)
+#else
+ # define Info(fmt, args ...)
+#endif
+
+
+AP_GPS_SBP2::AP_GPS_SBP2(AP_GPS &_gps, AP_GPS::GPS_State &_state,
+                       AP_HAL::UARTDriver *_port) :
+    AP_GPS_Backend(_gps, _state, _port)
+{
+    Info("SBP Driver Initialized");
+    parser_state.state = sbp_parser_state_t::WAITING;
+}
+
+// Process all bytes available from the stream
+//
+bool
+AP_GPS_SBP2::read(void)
+{
+    //Invariant: Calling this function processes *all* data current in the UART buffer.
+    //
+    //IMPORTANT NOTICE: This function is NOT CALLED for several seconds
+    // during arming. That should not cause the driver to die. Process *all* waiting messages
+
+    _sbp_process();
+    return _attempt_state_update();
+}
+
+void
+AP_GPS_SBP2::inject_data(const uint8_t *data, uint16_t len)
+{
+    if (port->txspace() > len) {
+        last_injected_data_ms = AP_HAL::millis();
+        port->write(data, len);
+    } else {
+        Debug("PIKSI: Not enough TXSPACE");
+    }
+}
+
+//This attempts to reads all SBP messages from the incoming port.
+//Returns true if a new message was read, false if we failed to read a message.
+void
+AP_GPS_SBP2::_sbp_process()
+{
+    int nleft = port->available(); 
+    while (nleft > 0) {
+        nleft--;
```

Comment:
uint32_t please.

Nitpicking: not required, but more compact code would be `while (nleft-- > 0)`, it is just a matter of style and we don't enforce any on this.